### PR TITLE
fix: endless retry if namespace doesn't exist

### LINKED ISSUE #816: bug: namespace update event will be retried unconditionally even if it was deleted
### Issue description

Namespace  events will be retried endlessly even if it's expired, for instance, an update event should be mark as expired if a new delete event is handled.

```
2021-12-29T10:12:04+08:00       warn    ingress/namespace.go:121        sync namespace info failed, will retry  {"namespace": "cp-387818008396956540", "error": "namespaces \"cp-387818008396956540\" not found"}
2021-12-29T10:12:04+08:00       warn    ingress/namespace.go:121        sync namespace info failed, will retry  {"namespace": "cp-387634093115835119", "error": "namespaces \"cp-387634093115835119\" not found"}
```

### Environment

- your apisix-ingress-controller version (output of apisix-ingress-controller version --long): 

```
2021-12-29T10:14:35+08:00       info    ingress/ingress.go:114  version:
Version: 1.3.0
Git SHA: no-git-module
Go Version: go1.13.8
Building OS/Arch: linux/amd64
Running OS/Arch: linux/amd64
```

- your Kubernetes cluster version (output of kubectl version): 

```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.5", GitCommit:"6b1d87acf3c8253c123756b9e61dac642678305f", GitTreeState:"archive", BuildDate:"2021-03-30T00:00:00Z", GoVersion:"go1.16", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.2-eks-06eac09", GitCommit:"5f6d83fe4cb7febb5f4f4e39b3b2b64ebbbe3e97", GitTreeState:"clean", BuildDate:"2021-09-13T14:20:15Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
```

- if you run apisix-ingress-controller in Bare-metal environment, also show your OS version (uname -a):

### Minimal test code / Steps to reproduce

Nope.

### Actual result

Stale events should be ignored.

### Error log

2021-12-29T10:12:04+08:00       warn    ingress/namespace.go:121        sync namespace info failed, will retry  {"namespace": "cp-387818008396956540", "error": "namespaces \"cp-387818008396956540\" not found"}
2021-12-29T10:12:04+08:00       warn    ingress/namespace.go:121        sync namespace info failed, will retry  {"namespace": "cp-387634093115835119", "error": "namespaces \"cp-387634093115835119\" not found"}

### Expected result

_No response_

### PR DESCRIPTION
**Pull request by lingsamuel**
Signed-off-by: Ling Samuel <lingsamuelgrace@gmail.com>

### Type of change:

Fixes #816 

- [x] Bugfix

### What this PR does / why we need it:

I couldn't find a stable way to reproduce this issue without changing the code, so there is no e2e test.
The only case this PR addresses is the error not found. Please add other cases if necessary.

If anyone else wants to reproduce this issue, please postpone the Create/Update events (example: wrap the function in a goroutine and add a time.Sleep at the beginning).

### CODE REVIEW

**Review by tao12345666333 — APPROVED** [review]
LGTM

**Comment by tao12345666333** [comment]
ping @gxthrj @tokers  PTAL, thanks!

File: pkg/ingress/namespace.go [hunk 1]
```diff
@@ -100,24 +104,32 @@ func (c *namespaceController) sync(ctx context.Context, ev *types.Event) error {
...
 
 func (c *namespaceController) handleSyncErr(event *types.Event, err error) {
 	name := event.Object.(string)
 	if err != nil {
+		if k8serrors.IsNotFound(err) && event.Type != types.EventDelete {
+			log.Infow("sync namespace but not found, ignore",
+				zap.String("event_type", event.Type.String()),
+				zap.String("namespace", event.Object.(string)),
+			)
+			c.workqueue.Forget(event)
```
Comment (gxthrj) [inline pkg/ingress/namespace.go:130]: So here is the point :)

**Review by gxthrj — APPROVED** [review]
LGTM

**Pull request merged**